lib/identity_cache/query_api.rb

```diff
@@ -246,96 +272,102 @@ def find_batch(ids)
         ids.map{ |id| records_by_id[id] }
       end
 
-      def prefetch_associations(associations, records)
-        associations = hashify_includes_structure(associations)
+      def fetch_embedded_associations(records)
+        associations = embedded_associations
+        return true if associations.empty?
 
-        associations.each do |association, sub_associations|
-          case
-          when details = cached_has_manys[association]
+        return false unless primary_cache_index_enabled
 
-            if details[:embed] == true
-              child_records = records.map(&details[:cached_accessor_name].to_sym).flatten
+        records_by_id = records.index_by(&:id)
+        cached_records = fetch_multi(records.map(&:id))
+
+        associations.each_value do |options|
+          cached_records.each do |cached_record|
+            record = records_by_id.fetch(cached_record.id)
+            if options[:embed] == :ids
+              cached_association = cached_record.public_send(options.fetch(:cached_ids_name))
+              record.instance_variable_set(:"@#{options.fetch(:ids_variable_name)}", cached_association)
             else
-              ids_to_parent_record = records.each_with_object({}) do |record, hash|
-                child_ids = record.send(details[:cached_ids_name])
-                child_ids.each do |child_id|
-                  hash[child_id] = record
-                end
-              end
+              cached_association = cached_record.public_send(options.fetch(:cached_accessor_name))
+              record.instance_variable_set(:"@#{options.fetch(:records_variable_name)}", IdentityCache.map_cached_nil_for(cached_association))
+            end
+          end
+        end
+        true
+      end
 
-              parent_record_to_child_records = Hash.new { |h, k| h[k] = [] }
-              child_records = details[:association_reflection].klass.fetch_multi(*ids_to_parent_record.keys)
-              child_records.each do |child_record|
-                parent_record = ids_to_parent_record[child_record.id]
-                parent_record_to_child_records[parent_record] << child_record
-              end
+      def prefetch_embedded_association(records, association, details)
+        first_record = records.first
+        return if first_record.association(association).loaded?
+        iv_name_key = details[:embed] == true ? :records_variable_name : :ids_variable_name
+        return if first_record.instance_variable_get(:"@#{details[iv_name_key]}")
+        fetch_embedded_associations(records)
+      end
 
-              parent_record_to_child_records.each do |parent_record, child_records|
-                parent_record.send(details[:prepopulate_method_name], child_records)
+      def prefetch_one_association(association, records)
```

Comment:
Oh, I see when reviewing with `?w=0` that this method actually didn't really change, just got unindented